web/packages/teleport/src/DesktopSession/KeyboardHandler.tsx

```diff
@@ -0,0 +1,271 @@
+    if (this.syncBeforeNextKey === true) {
+      cli.sendSyncKeys(this.getSyncKeys(e));
+      this.syncBeforeNextKey = false;
+    }
+  }
+
+  private getSyncKeys = (e: KeyboardEvent): SyncKeys => {
+    return {
+      scrollLockState: this.getModifierState(e, 'ScrollLock'),
+      numLockState: this.getModifierState(e, 'NumLock'),
+      capsLockState: this.getModifierState(e, 'CapsLock'),
+      kanaLockState: ButtonState.UP, // KanaLock is not supported, see https://www.w3.org/TR/uievents-key/#keys-modifier
+    };
+  };
+
+  /**
+   * Returns the ButtonState corresponding to the given `keyArg`.
+   *
+   * @param e The `KeyboardEvent`
+   * @param keyArg The key to check the state of. Valid values can be found [here](https://www.w3.org/TR/uievents-key/#keys-modifier)
+   */
+  private getModifierState = (
+    e: KeyboardEvent,
+    keyArg: string
+  ): ButtonState => {
+    return e.getModifierState(keyArg) ? ButtonState.DOWN : ButtonState.UP;
+  };
+
+  /**
+   * Called before every keydown or keyup event. This witholds the
+   * keys for which we want to see what the next event is before
+   * sending them on to the server, and sets up delayed up events.
+   *
+   * Returns true if the event was handled, false otherwise.
+   */
+  private handleWithholdingAndDelay(
+    cli: TdpClient,
+    e: KeyboardEvent,
+    state: ButtonState
+  ): boolean {
+    if (this.isWitholdableOrDelayeable(e) && state === ButtonState.DOWN) {
+      // Unlikely, but theoretically possible. In order to ensure correctness,
+      // we clear any delayed up event for this key and handle it immediately.
+      const timeout = this.delayedUp.get(e.code);
+      if (timeout) {
+        clearTimeout(timeout);
+        this.finishHandlingKeyboardEvent(cli, e, ButtonState.UP);
+      }
+
+      // Then we set the key down to be withheld until the next keydown or keyup,
+      // or to never be sent if this cache is cleared onfocusout.
+      this.withheldDown.set(e.code, true);
+
+      return true;
+    } else if (this.isWitholdableOrDelayeable(e) && state === ButtonState.UP) {
+      // If we receive a delayed up event for a key that was already delayed,
+      // we log a warning. This should never happen, because we can only get an
+      // up event after a down, and we ensure the up cache is cleared when we
+      // handle a down event.
+      if (this.delayedUp.has(e.code)) {
+        // eslint-disable-next-line no-console
+        console.warn(
+          'Received a delayed up event for a key that was already delayed. This should not happen.'
+        );
+      }
+
+      const timeout = setTimeout(() => {
+        this.finishHandlingKeyboardEvent(cli, e, ButtonState.UP);
+      }, 10 /* ms */);
+
+      // And add the timeout to the cache.
+      this.delayedUp.set(e.code, timeout);
+
+      return true;
+    }
+
+    return false;
+  }
+
+  private isWitholdableOrDelayeable(e: KeyboardEvent): boolean {
```

Comment:
```suggestion
  private isWitholdableOrDelayable(e: KeyboardEvent): boolean {
```

I think?

Reply:
Perhaps even `isWithholdable…`? https://en.wiktionary.org/wiki/withholdable

Reply:
Becomes irrelevant in 44595889eba30982176bf283d415879cf35e169d